src/BidirectionalFlatList.tsx

```diff
@@ -69,185 +48,469 @@ export type Props<T> = Omit<
+        if (!vData) {
+          return null;
+        }
 
-      if (HeaderLoadingIndicator) {
-        return <HeaderLoadingIndicator />;
-      }
+        const indexInData = inverted
+          ? vData.length - 1 - (index + Math.abs(firstItemIndex.current))
+          : index + Math.abs(firstItemIndex.current);
 
-      return (
-        <View style={styles.indicatorContainer}>
-          <ActivityIndicator size={'small'} color={activityIndicatorColor} />
-        </View>
-      );
-    };
+        return (
+          <>
+            {renderItem({
+              index,
+              item: vData[indexInData],
+              separators: {
+                highlight: () => null,
+                unhighlight: () => null,
+                updateProps: () => null,
+              },
+            })}
+            {indexInData !== vData.length && !!ItemSeparatorComponent && (
+              <ItemSeparatorComponent />
+            )}
+          </>
+        );
+      },
+      [vData?.length]
+    );
 
-    const renderFooterLoadingIndicator = () => {
-      if (!showDefaultLoadingIndicators) {
-        if (ListFooterComponent) {
-          return <ListFooterComponent />;
-        } else {
-          return null;
+    useEffect(() => {
+      const updateVirtuoso = () => {
+        if (!data?.length) {
+          return;
+        }
+
+        let nextFirstItemIndex = firstItemIndex.current;
+        if (prependingItems.current) {
+          nextFirstItemIndex =
+            firstItemIndex.current - (data.length - previousDataLength.current);
         }
-      }
 
-      if (!onEndReachedInProgress) return null;
+        appendingItems.current = false;
+        prependingItems.current = false;
+        previousDataLength.current = data.length;
 
-      if (FooterLoadingIndicator) {
-        return <FooterLoadingIndicator />;
-      }
+        firstItemIndex.current = nextFirstItemIndex;
+        setVDate(() => data);
+
+        if (
+          enableAutoscrollToTop &&
+          offsetFromBottom.current < autoscrollToTopThreshold
+        ) {
+          setTimeout(() => {
+            virtuosoRef.current?.scrollToIndex({
+              behavior: 'smooth',
+              index: data.length,
+            });
+          }, 0);
+        }
+      };
 
+      updateVirtuoso();
+    }, [enableAutoscrollToTop, autoscrollToTopThreshold, data, setVDate]);
+
+    useImperativeHandle(
+      ref,
+      // @ts-ignore
+      () => ({
+        flashScrollIndicators: () => null,
+        getNativeScrollRef: () => null,
+        getScrollableNode: () => null,
+        getScrollResponder: () => null,
+        recordInteraction: () => null,
+        scrollToEnd: (params = { animated: true }) => {
+          const { animated } = params;
+          if (!vData?.length) return;
+
+          if (inverted) {
+            virtuosoRef.current?.scrollToIndex({
+              behavior: animated ? 'smooth' : 'auto',
+              index: 0,
+            });
+          } else {
```

Comment:
This can be switched to 

```js
const index = inverted ? 0 : vData.length - 1;
virtuosoRef.current?.scrollToIndex({
  behavior: animated ? 'smooth' : 'auto',
  index,           
});
```

Reply:
Thanks , I will check it out. Planning to publish beta on this next week :) Please let me know if you find more things.
PR needs some testing and some cleanup which will happen early next week :)